[Recorder] Readme (#21950)
Azure/azure-sdk-for-js: sdk/test-utils/recorder/README.md

The test-recorder provides the `Recorder` class that deals with recording and playing back the network requests, depending on the value assigned to the `TEST_MODE` environment variable.

- If `TEST_MODE` equals to `record`, it will automatically store network requests in a plain text file in the folder `recordings` at the root of your repository (which for our example case is the root of the `sdk/tables/data-tables` project).
- This package assumes that the tests in the sdk are leveraging
  [mocha](https://mochajs.org/) and [rollup](https://rollupjs.org/guide/en/)
  (and [karma](https://karma-runner.github.io/latest/index.html) test runner for browser tests) as suggested by the [template](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/template/template) package in the repo.

#### package.json scripts

For the unified recorder client library to work, the [test proxy server] must be active while you are running your tests. Helpers have been added to the `dev-tool` package which manage starting and stopping the test proxy server before and after your tests are run.

> The following commands run the tests with the default configs, and concurrently starts(runs) the test-proxy tool in a detached process the background in record/playback modes if it is not already active. Additionally, more options can be passeed to override the default configs.

Author: ```suggestion
The following commands run the tests with the default configs and concurrently starts(runs) the test-proxy tool in a detached process in the background in record/playback modes if it is not already active. Additionally, more options can be passed to override the default configs.
```

```js
const client = new AnyCoreV2Client(/** args **/, recorder.configureClientOptions(/** client options **/));
```

- Modifying the client options(that you'd otherwise provide) with `recorder.configureClientOptions()` helps us with adding the recording policy to the additionalPolicies in the client options.
- Adding the policy is important for redirecting the requests to the proxy tool instead of directly going to the service.

  _Note: If your client is relying on `@azure/core-http` instead of the core-v2 libraries(i.e., `@azure/core-client` and `@azure/core-rest-pipeline`), please use `recorder.configureClientOptionsCoreV1()` instead of `recorder.configureClientOptions()`._

Once instantiated the recorder, you're expected to start the recorder using the `recorder.start()` method with the appropriate recorder options.

```js
await recorder.start(/** recorderOptions go here **/);
```

- Recorder options will typically contain the environment setup needed for the `playback` mode, and the sanitizers that help with masking the sensitive information in the recordings, more on the recorder options below.

> After the start call is made, any requests that are made using the `client (AnyCoreV2Client)` will be redirected to the test-proxy tool before they reach the service, the requests and responses will be recorded and saved when `recorder.stop()` is called in `record` mode.

Author: Good point, edited a little, let me know if that sounds better

Likewise, in `playback` mode, the saved responses are utilized by the test-proxy tool when the requests are redirected to it instead of reaching the service.

```js
await recorder.stop();
```

- Call this method to ping the test-proxy tool with a stop request, this helps to stop recording saving the recording file in record mode.

_Note: Instantiating the recorder, starting and stopping the recorder, have no effects in the `live` mode(`TEST_MODE=live`), means the redirection to the test-proxy tool doesn't happen and the requests would reach the services as usual._

#### Recorder#variable()

> To handle the dynamic/generated values for testing that are created as part of the tests, to make sure the requests in the `playback` mode match the ones in the `record` mode, you can leverage the `Recorder#variable` function.

Author: Added some docs, please check

> To handle the dynamic/generated values for testing that are created as part of the tests, to make sure the requests in the `playback` mode match the ones in the `record` mode, you can leverage the `Recorder#variable` function.

Reviewer: Thanks, looks good.

#### karma.conf - for the browser tests

When running browser tests, the recorder relies on an environment variable to determine where to save the recordings. Add this snippet to your `karma.conf.js`:

```ts
const { relativeRecordingsPath } = require("@azure-tools/test-recorder");

process.env.RECORDINGS_RELATIVE_PATH = relativeRecordingsPath();
```

And then, again in `karma.conf.js`, add the variable to the list of environment variables:

```ts
module.exports = function (config) {
  config.set({
    /* ... */

    envPreprocessor: [
      ,
      /* ... */ "RECORDINGS_RELATIVE_PATH", // Add this!
    ],

    /* ... */
  });
};
```

## Examples

### How to record

To record your tests,

- make sure to set the environment variable `TEST_MODE` to `record`

  ```sh
  # Windows
  set TEST_MODE=record

  # Linux / Mac
  export TEST_MODE=record
  ```

- then in your code, instantiate the `Recorder`
- call `#start()` function with the recorder options
- modify the client options using the `#configureClientOptions()` method
- then make calls with your client as needed
- call `#stop()` function to save the recording in a file

> In the following example, we'll use the recorder with the client from `@azure/data-tables`:

Author: sure ❤️

If you launch the `recorder` in record mode with some of these changes (and given that you activate the recorder on `beforeEach`), only the files that relate to the changed tests will be updated. Skipped tests won't update their recordings.

This way, you can focus on fixing a specific set of tests with `.only`, then remove all the `.only` calls and trust that the playback will keep confirming that the unaffected tests are fine and green.

You can also skip specific tests with the following.

```js
import { isLiveMode } from "@azure-tools/test-recorder";

it("test-title", function () {
  // isPlaybackMode() and isRecordMode() methods are also available from recorder.
  if (!isLiveMode()) this.skip(); // This skips the test in record and playback modes
  // Test goes here...
  // ...
});
```

### Supporting parallelism

A common issue while running integration tests is that, sometimes two individuals or machines might try to run the same set of tests against the same resource.

> This is not directly related to the `@azure-tools/test-recorder` package, but if you're getting into issues because of concurrent conflicting requests, we strongly suggest using randomly generated strings as prefixes or suffixes for the resources you create.

Reviewer: Might be worth linking to the section on `Recorder.variable` in this subsection since you'd need to use that if generating random strings etc.

Author: This was also a section I lifted from recorder v1.
You raise a great point. 💯

| :------------------------- | :--------------------------------------------------------------------------------------------------------------- |
| `unit-test:browser`        | `dev-tool run test:browser`                                                                                      |
| `unit-test:node`           | `dev-tool run test:node-ts-input -- --timeout 1200000 --exclude 'test/**/browser/*.spec.ts' 'test/**/*.spec.ts'` |
| `integration-test:browser` | `dev-tool run test:browser`                                                                                      |
| `integration-test:node`    | `dev-tool run test:node-js-input -- --timeout 5000000 'dist-esm/test/**/*.spec.js'`                              |

Note the difference between the dev-tool `node-ts-input` and `node-js-input` commands:

- `node-ts-input` runs the tests using `ts-node`, without code coverage.
- `node-js-input` runs the tests using the built JavaScript output, and generates coverage reporting using `nyc`.

Read more at [dev-tool commands #usage](https://github.com/Azure/azure-sdk-for-js/blob/main/common/tools/dev-tool/README.md#usage)

#### Prerequisites

> - [Docker](https://docker.com/) is required, as the [test proxy server](https://github.com/Azure/azure-sdk-tools/tree/main/tools/test-proxy) is run in a container during testing. When running the tests, ensure the Docker daemon is running and you have permission to use it. For WSL 2, running `sudo service docker start` and `sudo usermod -aG docker $USER` should be sufficient.

Reviewer 2: I think it's worth adding something here with the installation link to Docker for windows. When I was reading this, I assumed I needed to use WSL.

Author: updated :)

> - [Docker](https://docker.com/) is required, as the [test proxy server](https://github.com/Azure/azure-sdk-tools/tree/main/tools/test-proxy) is run in a container during testing. When running the tests, ensure the Docker daemon is running and you have permission to use it. For WSL 2, running `sudo service docker start` and `sudo usermod -aG docker $USER` should be sufficient.

Author: ```suggestion
- [Docker](https://docker.com/) is required, as the [test proxy server](https://github.com/Azure/azure-sdk-tools/tree/main/tools/test-proxy) is run in a container during testing. When running the tests, ensure the Docker daemon is running and you have permission to use it. 

Check [docker.com/get-started](https://www.docker.com/get-started/) to download and install docker desktop on your machine.

For WSL 2, running `sudo service docker start` and `sudo usermod -aG docker $USER` should be sufficient.
```

## Getting started

We're about to go through how to set up your project to use the `@azure-tools/test-recorder` package.

**_Note: If you're new to this repository, follow the [ContributingGuide.md](https://github.com/Azure/azure-sdk-for-js/blob/main/CONTRIBUTING.md) to learn how to setup/build the repo and to create/test an SDK in the [Azure/azure-sdk-for-js](https://github.com/Azure/azure-sdk-for-js) repository._**

### Installing the package

To install the `@azure-tools/test-recorder` package, you can use the following command.

```bash
rush add -p @azure-tools/test-recorder@^2.0.0 --dev
```

From this point forward, we'll assume that you're developing (perhaps contributing!) to one of the azure-sdk-for-js's libraries. So, your next step is to change directory to the path relevant to your project. Let's say you want to add the `@azure-tools/test-recorder` package to `@azure/data-tables` (it already uses test-recorder, but bear with us), you'll be doing the following:

```bash

> cd sdk/tables/data-tables

Author: Great catch

> ```bash
> cd sdk/tables/data-tables
> ```

Author: ```suggestion
```bash
cd sdk/tables/data-tables
rush add -p @azure-tools/test-recorder@^2.0.0 --dev
```
```

- To **playback** means to intercept any HTTP request and to respond it with the stored response of a previously recorded matching request.
- **Sensitive information** means content that should not be shared publicly. Content like passwords, unique identifiers or personal information should be cleaned up from the recordings. Some functionality is provided to fix this problem. You can read more at [securing sensitive data](#securing-sensitive-data).

## Getting started

We're about to go through how to set up your project to use the `@azure-tools/test-recorder` package.

This document assumes familiarity with [git](https://git-scm.com) and [rush](https://rushjs.io). You can read more about how we use rush in the following links:

- Rush used for [Project Orchestration](https://github.com/sadasant/azure-sdk-for-js/blob/master/CONTRIBUTING.md#project-orchestration).
- [Rush for NPM users](https://github.com/sadasant/azure-sdk-for-js/blob/master/CONTRIBUTING.md#rush-for-npm-users).

> ### Installing the package

Reviewer 3: I am confused by this section because it explains how to close and build the repo, things I would expect to be in the contributing guide but not in an internal package readme. Could we instead just put a command to add the package to package.json, e.g. `rush add -p @azure-tools/test-recorder@^2.0.0 --dev`? the document can then refer to the contributing guide for instructions on how to link and build the deps.

Author: Great point, I'll update this.
Retained this from recorder v1, but I would agree with what you said.

Author: Updated some things, is it nicer now?

```json
{
  // ... your package.json properties
  "devDependencies": {
    // ... your devDependencies
    "@azure-tools/test-credential": "^1.0.0", // If you are using `@azure/identity` in your tests
    "@azure-tools/test-recorder": "^2.0.0"
    // ... more of your devDependencies
  }
  // ... more of your package.json properties
}
```

After that, we recommend you to update rush and install the dependencies again, as follows:

```bash
rush update && rush install
```

And you're ready! Now you can use the common recorder in your code, as shown below:

```typescript
import { Recorder } from "@azure-tools/test-recorder";
```

Or, if you know what you want to import, you can also do the following:

```typescript
import { Recorder, RecorderStartOptions, env, SanitizerOptions } from "@azure-tools/test-recorder";
```

### Configuring your project

Having the recorder as a devDependency means that you'll be able to start recording tests right away by using the `Recorder` class.

The test-recorder provides the `Recorder` class that deals with recording and playing back the network requests, depending on the value assigned to the `TEST_MODE` environment variable.

> - If `TEST_MODE` equals to `record`, it will automatically store network requests in a plain text file in the folder `recordings` at the root of your repository (which for our example case is the root of the `sdk/tables/data-tables` project).

Author: Also retained what Daniel wrote for recorder v1.
Your wording sounds better, agreed.

Interactions with the test-proxy tool vary based on what the `TEST_MODE` environment is.
|TEST_MODE |What? |
| :--------- | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `record` | Stores network requests with the help of test-proxy tool in a plain text file in the folder `recordings` at the root of your repository (example: root of the `sdk/tables/data-tables` project) |
| `playback` | Stored requests/responses are utilized by the test-proxy tool when the requests are redirected to it instead of reaching the service |  
| `live` | Recorder and its methods are no-ops here, requests directly reach the service instead of being redirected at the test-proxy tool layer |

### Using the `Recorder`

Inside a mocha test (either in the `beforeEach` or in the test body itself), you will need to instantiate the `Recorder` as below to leverage its functionalities.

```js
let recorder: Recorder;

beforeEach(async function (this: Mocha.Context) {
  recorder = new Recorder(this.currentTest);
});
```

> Adding the recorder policies to your client, whose requests are supposed to be recorded.
> 
> ```js
> const client = new AnyCoreV2Client(/** args **/, recorder.configureClientOptions(/** client options **/));
> ```
> 
> - Modifying the client options(that you'd otherwise provide) with `recorder.configureClientOptions()` helps us with adding the recording policy to the additionalPolicies in the client options.
> - Adding the policy is important to redirect the requests to the proxy tool instead of directly going to the service.

Reviewer 3: nit: simplification
```suggestion
The tested client needs to install the recording policy that redirects requests to the recording proxy first before they go to the service. This is done by simply passing the client options bag through the `recorder.configureClientOptions` helper:

```js
const client = new AnyCoreV2Client(/** args **/, recorder.configureClientOptions(/** client options **/));
```
```

  _Note: If your client relies on `@azure/core-http` instead of the core-v2 libraries(i.e., `@azure/core-client` and `@azure/core-rest-pipeline`), please use `recorder.configureClientOptionsCoreV1()` instead of `recorder.configureClientOptions()`._

> Once the recorder has been instantiated, you must start the recorder using the `recorder.start()` method with the appropriate recorder options.

Reviewer 3: nit: the language could be simplified a bit
```suggestion
Recording can start using the `recorder.start()` method.
```
